kube-core/src/labels.rs

```diff
@@ -0,0 +1,552 @@
+    In(String, BTreeSet<String>),
+
+    /// Key does not exists or not in set:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::NotIn("foo".into(), ["bar".into(), "baz".into()].into()).to_string();
+    /// assert_eq!(exp, "foo notin (bar,baz)");
+    /// let exp = Expression::NotIn("foo".into(), vec!["bar".into(), "baz".into()].into_iter().collect()).to_string();
+    /// assert_eq!(exp, "foo notin (bar,baz)");
+    /// ```
+    NotIn(String, BTreeSet<String>),
+
+    /// Key exists and is equal:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::Equal("foo".into(), "bar".into()).to_string();
+    /// assert_eq!(exp, "foo=bar")
+    /// ```
+    Equal(String, String),
+
+    /// Key does not exists or is not equal:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::NotEqual("foo".into(), "bar".into()).to_string();
+    /// assert_eq!(exp, "foo!=bar")
+    /// ```
+    NotEqual(String, String),
+
+    /// Key exists:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::Exists("foo".into()).to_string();
+    /// assert_eq!(exp, "foo")
+    /// ```
+    Exists(String),
+
+    /// Key does not exist:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::DoesNotExist("foo".into()).to_string();
+    /// assert_eq!(exp, "!foo")
+    /// ```
+    DoesNotExist(String),
+
+    /// Invalid combination. Always evaluates to false.
+    Invalid,
+}
+
+/// Perform selection on a list of expressions
+#[derive(Clone, Debug, Eq, PartialEq, Default, Deserialize, Serialize)]
+pub struct Selector(Expressions);
+
+impl Selector {
+    /// Create a selector from a vector of expressions
+    fn from_expressions(exprs: Expressions) -> Self {
+        Self(exprs)
+    }
+
+    /// Create a selector from a map of key=value label matches
+    fn from_map(map: BTreeMap<String, String>) -> Self {
+        Self(map.into_iter().map(|(k, v)| Expression::Equal(k, v)).collect())
+    }
+
+    /// Indicates whether this label selector matches all pods
+    pub fn selects_all(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Extend the list of expressions for the selector
+    ///
+    /// ```
+    /// use kube_core::Selector;
+    /// use kube_core::Expression;
+    /// use k8s_openapi::apimachinery::pkg::apis::meta::v1::LabelSelector;
+    ///
+    /// let label_selector: Selector = LabelSelector::default().into();
+    /// let mut selector = Selector::default();
+    /// selector.extend(Expression::Equal("environment".into(), "production".into()));
+    /// selector.extend([Expression::Exists("bar".into()), Expression::Exists("foo".into())].into_iter());
+    /// selector.extend(label_selector);
+    /// ```
+    pub fn extend(&mut self, exprs: impl IntoIterator<Item = Expression>) -> &Self {
+        self.0.extend(exprs);
+        self
```

Comment:
Why does this return `&Self`? You are accepting a `&mut self`; if you want to do a fluent setter, shouldn't this return `&mut` so that you can extend more?

Reply:
I see this has been updated now in https://github.com/kube-rs/kube/pull/1539/commits/fa7b856a49ac785a82ac27fddad7361709c8239b so resolving this.

Reply:
> do you even need a fluent setter to begin with instead of from_expressions?

`from_expressions` only allows setting from one input type, so the extend here is a nice way to merge two input sources imo.